### ISSUE TITLE
[Suggestion] Export item to clipboard

### ISSUE THREAD

**Issue by OSwoboda** [xyzz/acquisition#448]
Would be nice if an item could be exported to clipboard to be able to import it to path of building for example.

**Comment by DaneelTrevize** [comment]
Anyone know the current intended use of the Tooltip Text tab?  
I find it's set to Qt::FocusPolicy NoFocus, so it can't even be used as a place to click, Ctrl+A, Ctrl+C, paste into PoB's Create custom... item form and massage from there.

Changing it to StrongFocus enables clicking, but it would need a few tweaks to be able to be imported without edits:
Split the name and base item type onto different lines;
Ditch everything else before the mods lists; remove extra newlines from badly handled new hybrid mods such +armour/+ES;

The resulting text imports, and generates the following central attributes (for a body armour):
`Quality: 20
Sockets: ...
LevelReq: ...
Implicits: 0`

As for UX, if this text tooltip tab were to be repurposed for this use-case, I'm thinking a button to "Export to Clipboard in Path of Building format" might be better, and then I'm wondering why the last tab is Hide..? It doesn't hide the whole tooltip pane (with socketed image above and stash map below). And the Tooltip first tab has an "Upload to Imgur" button, so perhaps our Export and hide/collapse UI elements would be best placed below that and on the collapsing area?

**Comment by xyzz** [comment]
Yeah, the purpose is so that it displays same information as the fancy tooltip view but takes less space. The "hide" tab is to hide the tooltip, making the right pane take even less space.

I think instead of changing the tooltip a better approach would be to add a new button to the right column, "copy to clipboard" (and maybe support shortcuts like ctrl+c), which copies pob-compatible data (so just like you described, but instead of changing the text tooltip it's completely separate).

**Comment by DaneelTrevize** [comment]
> The "hide" tab is to hide the tooltip, making the right pane take even less space.

Ok, but I don't experience that:  
[Before](https://i.imgur.com/VOLDAlG.png), with the area sized just smaller than the element, triggering the scroll bar.  
[After](https://i.imgur.com/x2owMT5.png), with no change to the invisible width and now wasting space.

I think that displaying the PoB-format text that it would generate would help people understand what a button would do, similar to how the fancy tooltip does above the Imgur button. I'm no UX expert but this seems reasonable.

**Comment by xyzz** [comment]
Yeah, unfortunately the layout code leaves much to be desired, I think it still works in some cases when you have a longer item name or mods. If you're worried about discoverability, you can just name the button "copy to pob".

**Comment by DaneelTrevize** [comment]
Note to self: PoB creates player-input custom items via a call to the method ItemsTabClass:CreateDisplayItemFromRaw(), which eventually calls ItemClass:ParseRaw(), as found [here](https://github.com/Openarl/PathOfBuilding/blob/46b9d532e2816c49f8610787a249bed1787074e4/Classes/Item.lua#L24).  
This seems to provide the strings and regex to target generating matches for, but not an overall data structure or schema to validate against.

**Comment by DaneelTrevize** [comment]
To generate the PoB format for a given item, we first need to know if it's a wearable item (including flasks).

This lead down a rabbit hole where flasks aren't correctly categories by Acquisition, because their 2D art isn't generated under the same CDN path format as other items, possibly because of the POE website displaying flask charges within the image. Which leads me to question why we aren't using the "category" object in an item's JSON, rather than regex over the icon path..?  
Incomplete [proof of concept](https://i.imgur.com/ds8vlJ2.png), but other than wanting a mapping of backend to pretty strings, and a few conditional statements for nesting certain subcategories as is done currently (i.e. not for gems & jewels, thought that's questionable), why not do it this way via the JSON objects hierarchy rather than CDN paths?

**Comment by xyzz** [comment]
It's using the path because when that feature was implemented the category field wasn't exposed via json; it's a good idea to change it now though.

**Comment by DaneelTrevize** [comment]
Proposed new implementation of categories, and thus Wearable, is available for review under PR #507 , so the work on the PoB export format itself can continue distinct from these preparatory steps.

**Comment by DaneelTrevize** [comment]
POB export functionality is found in PR #508

**Comment by OSwoboda** [comment]
I tested it and there is a major bug: Rare items are imported as uniques and magic items can't be imported because its tried to import them as rares.

**Comment by DaneelTrevize** [comment]
IIRC the PoB "Create custom..." dialog does not auto-select the correct rarity of item you're creating due to the mod list pasted, instead it just default to the last used one. Please confirm you're not just experiencing this aspect of their UI. And if so, perhaps raise an enhancement request [here](https://github.com/Openarl/PathOfBuilding/issues).

**Comment by OSwoboda** [comment]
I tested it with c&p from an ingame item:

> Rarity: Unique
> Presence of Chayula
> Onyx Amulet
> --------
> Requirements:
> Level: 60
> --------
> Item Level: 42
> --------
> +16 to all Attributes
> --------
> 30% increased Rarity of Items found
> +60% to Chaos Resistance
> Cannot be Stunned
> 20% of Maximum Life Converted to Energy Shield
> --------
> The dreamer stirs, the world trembles. 

and from acquisiton:

> Presence of Chayula
> Onyx Amulet
> Implicits: 1
> +16 to all Attributes
> 30% increased Rarity of Items found
> +60% to Chaos Resistance
> Cannot be Stunned
> 20% of Maximum Life Converted to Energy Shield

As you can see the information about rarity is missing. This would be ignoring the combobox in the "create custom..." dialog and also would support the option to create an item with just pasting it into the item tab and click add to build instead of "create custom" -> paste -> create -> add to build.

**Comment by DaneelTrevize** [comment]
Ok, quick testing indicates the "Rarity:" line can be added and immediately affects the PoB dialog's mouseover tooltip, but doesn't change the dropdown. Saving the item at this point will set the rarity, so it will work. Improvement might be added before Betrayal.

The reason this wasn't in the current version of Acquisition is because this feature was developed by finding the minimal version of what PoB presents back (when you subsequently edit a custom item) that works to (re)create the time. i.e. PoB can calculate the level req. itself from the base, so that's a redundant line.  
But perhaps instead this feature should be adjusted to generate the same clipboard contents as PoE does itself, for use with any other tools that accept that, rather than a PoB-specific minimised version?

**Comment by OSwoboda** [comment]
> But perhaps instead this feature should be adjusted to generate the same clipboard contents as PoE does itself, for use with any other tools that accept that, rather than a PoB-specific minimised version?

I think that would be the best general solution, using it for pob was just an example in my suggestion.

Thanks for listening to my concerns, looking forward to the improvement!
